File: Src/Visual Studio/SDK/C#/Eruru.ChatRobotRPC For Shared Project/WaitSystem.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace Eruru.ChatRobotRPC {

	class WaitSystem : IDisposable {

		public int MillisecondsTimeout {

			get {
				return _MillisecondsTimeout;
			}

			set {
				_MillisecondsTimeout = value;
			}

		}

		readonly Queue<Wait> WaitPool = new Queue<Wait> ();
		readonly List<Wait> Waits = new List<Wait> ();
		readonly object GetIDLock = new object ();
		readonly object WaitsLock = new object ();

		int ID;
		int _MillisecondsTimeout = 10 * 1000;

		public int GetID () {
			lock (GetIDLock) {
				if (ID > 100000) {
					ID = 0;
				}
				return ID++;
			}
		}

		public void Set (int id, string result) {
			lock (WaitsLock) {
				Wait wait = Waits.Find (item => item.ID == id);
				if (wait == null) {
					wait = NewWait (id);
					Waits.Add (wait);
				}
				wait.Result = result;
				wait.Signal = true;
				Monitor.PulseAll (WaitsLock);
				//throw new KeyNotFoundException (string.Format ("没有找到ID为{0}的Wait，Result：{1}", id, result));
			}
		}

		public string Get (int id) {
			lock (WaitsLock) {
				Wait wait = Waits.Find (item => item.ID == id);
				if (wait == null) {
					wait = NewWait (id);
					Waits.Add (wait);
				}
				while (true) {
					if (wait.Signal) {
						break;
					}
					if (!Monitor.Wait (WaitsLock, MillisecondsTimeout)) {
						throw new TimeoutException (string.Format ("ID为{0}的Wait等待超时", id));
					}
				}
				WaitPool.Enqueue (wait);
				Waits.Remove (wait);
				if (Waits.Count <= 0) {
					Monitor.PulseAll (WaitsLock);
				}
				return wait.Result;
			}
		}

		public void Dispose () {
			lock (WaitsLock) {
				ID = 0;
				while (Waits.Count > 0) {
					foreach (var wait in Waits) {
						wait.Signal = true;
					}
					Monitor.PulseAll (WaitsLock);
					if (!Monitor.Wait (WaitsLock, MillisecondsTimeout)) {
						throw new TimeoutException ();
					}
				}
			}
		}

		Wait NewWait (int id) {
			lock (WaitsLock) {
				Wait wait;
				if (WaitPool.Count == 0) {
					wait = new Wait ();
				} else {
					wait = WaitPool.Dequeue ();
					wait.Result = null;
					wait.Signal = false;
				}
				wait.ID = id;
				return wait;
			}
		}

		class Wait {

			public long ID;
			public string Result;
			public bool Signal;

		}

	}

}

File: Src/Intelij IDEA/SDK/Java/Eruru.ChatRobotRPC/src/main/java/org/eruru/chatrobotrpc/WaitSystem.java
package org.eruru.chatrobotrpc;

import javax.naming.NameNotFoundException;
import java.io.Closeable;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.TimeoutException;

class WaitSystem implements Closeable {

	private final Queue<Wait> waitPool = new LinkedList<> ();
	private final List<Wait> waits = new ArrayList<> ();
	private final Object getIDLock = new Object ();
	private final Object waitsLock = new Object ();

	private long millisecondsTimeout = 10 * 1000;
	private Long id = 0L;

	public long getMillisecondsTimeout () {
		return millisecondsTimeout;
	}

	public void setMillisecondsTimeout (long millisecondsTimeout) {
		this.millisecondsTimeout = millisecondsTimeout;
	}

	public long getID () {
		synchronized (getIDLock) {
			if (id == 100000) {
				id = 0L;
			}
			return id++;
		}
	}

	public void set (long id, String result) throws Exception {
		synchronized (waitsLock) {
			for (Wait wait : waits) {
				if (wait.id == id) {
					wait.result = result;
					wait.autoResetEvent.set ();
					return;
				}
			}
			throw new NameNotFoundException (String.format ("没有找到ID为%d的Wait，Result：%s", id, result));
		}
	}

	public String get (long id) throws InterruptedException, TimeoutException {
		Wait wait;
		synchronized (waitsLock) {
			wait = waitPool.size () == 0 ? new Wait () : waitPool.poll ();
			wait.id = id;
			wait.result = null;
			wait.autoResetEvent.reset ();
			waits.add (wait);
		}
		try {
			if (!wait.autoResetEvent.waitOne (millisecondsTimeout)) {
				throw new TimeoutException (String.format ("ID为%d的Wait等待超时", id));
			}
		} finally {
			synchronized (waitsLock) {
				waitPool.offer (wait);
				waits.remove (wait);
			}
		}
		return wait.result;
	}

	@Override
	public void close () {
		synchronized (waitsLock) {
			for (Wait wait : waits) {
				wait.autoResetEvent.close ();
				waitPool.offer (wait);
			}
			waits.clear ();
		}
	}

	static class Wait {

		public long id;
		public String result;
		public AutoResetEvent autoResetEvent = new AutoResetEvent (false);

	}

}